openedx_learning/apps/authoring/linking/models.py

```diff
@@ -0,0 +1,128 @@
+"""
+Contains models to store upstream-downstream links between publishable entities.
+"""
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from openedx_learning.lib.fields import (
+    case_insensitive_char_field,
+    immutable_uuid_field,
+    key_field,
+    manual_date_time_field,
+)
+
+from ..publishing.models import PublishableEntity
+
+__all__ = [
+    "PublishableEntityLink",
+    "LearningContextLinksStatus",
+    "LearningContextLinksStatusChoices"
+]
+
+
+class PublishableEntityLink(models.Model):
+    """
+    This represents link between any two publishable entities or link between publishable entity and a course
+    xblock. It helps in tracking relationship between xblocks imported from libraries and used in different courses.
+    """
```

Comment:
Question: Is the fact that this is living in openedx-learning distorting the design and making it harder than it needs to be? Because if this was in edx-platform, we'd be using UsageKey fields, CourseKeyFields, and using a join to CourseOverview for the context title, right?

Reply:
Except for usagekey and coursekey fields, there won't be much difference. All details about the xblock can be fetched from meilisearch index from the frontend, see this wip [code section](https://github.com/open-craft/frontend-app-course-authoring/blob/b58879605f14c0038c34f9803eb4857d636b5d10/src/course-libraries/CourseLibraries.tsx#L46-L53) as `CourseOverview` can only help us with course name, while we need block name, description, type, location etc.

Reply:
Why are we storing the "context title" anyways?

I am also wondering if edx-platform is a better place for this. Because Learning Core generally deals with Components, PublishableEntities, and LearningPackages, whereas these `____stream_usage_key` and `____stream_context_key` fields are more platform-specific workarounds for the fact that courses are not yet using Learning Core.

`usage_key` and `context_key` don't yet appear in this codebase, for example.

If we do want to keep this approach though, would it be crazy to have a `LegacyXBlock(PublishableEntityMixin)` that wraps a `UsageKey` field?

Reply:
In fact, I realize now the concepts of "upstream/downstream syncing" don't yet appear in Learning Core at all anyways. So it seems a bit odd if we have some APIs for tracking the links here, but no sync code, and the edx-platform codebase has APIs for syncing only.

We need to decide if this is a useful primitive that we want people to use to build things with, or if this is a layer on top of learning core that's more platform specific. There's a future world where everything is in Learning Core, and this model can be much simpler - ForeignKey(PublishableEntity) on both sides. I am wondering if we should wait until then before moving it into Learning Core. But i don't really have a strong opinion on this.

Reply:
> Why are we storing the "context title" anyways?

Displaying list of courses using the library block in library authoring page. We can however remove it and fetch the title (and other details if required) via meilisearch index.

Reply:
>  There's a future world where everything is in Learning Core, and this model can be much simpler - ForeignKey(PublishableEntity) on both sides. I am wondering if we should wait until then before moving it into Learning Core.

This seems like a good option, we will be in a much better position then.

Reply:
I agree with @bradenmacdonald on this. Apologies if I was the one that suggested it be here in the first place, but after seeing how it needs to be implemented, I think it's better in edx-platform for now.

Reply:
@ormsbee @bradenmacdonald Done, moved this to edx-platform. We can close this PR now.